### PR TITLE
[GPU] Fuse Select (where) attention masks into SDPA via existing SDPA fusion

### PR DESCRIPTION
**Pull request by clee30**
Some models decompose masked attention with a where-style mask: MatMul(Q,K) -> Multiply(scale) -> Select(mask, scores, neg_inf) -> Softmax -> MatMul(probs, V).   The common ov::pass::SDPAFusion only matches an additive (Add) mask, so these layers stay decomposed on GPU and miss the efficient ocl::sdpa kernels.

Add matcher pass ov::intel_gpu::SDPASelectMaskFusion that rewrites the where-mask into the additive form the common SDPAFusion already understands:

    Select(mask, scores, neg_inf) == scores + Select(mask, 0, neg_inf)

valid because the following Softmax normalizes the masked-out (-inf) entries to 0. It is registered before CommonOptimizations so the emitted Add(scores, mask) -> Softmax -> MatMul pattern is fused by the common ov::pass::SDPAFusion.

### Tickets:
 - CVS-174294

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+        // The rewrite is only equivalent once normalized by a following Softmax.
+        if (!feeds_softmax(pm.at(select_m)))
+            return false;
+
+        // additive_mask = Select(mask, 0, neg_inf); result = scores + additive_mask.
+        auto zero = v0::Constant::create(et, ov::Shape{}, {0.0f});
+        auto neg_inf_new = v0::Constant::create(et, ov::Shape{}, {neg_inf_val});
+        auto add_mask = std::make_shared<v1::Select>(cond_out, zero, neg_inf_new);
+        auto add = std::make_shared<v1::Add>(scores_out, add_mask);
```
[the remark this answers is not in the document]
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp:104]: add the fix.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+namespace {
+bool feeds_softmax(const ov::Output<ov::Node>& out) {
+    for (const auto& in : out.get_target_inputs()) {
+        auto* consumer = in.get_node();
+        if (ov::is_type<v8::Softmax>(consumer))
+            return true;
+        if (ov::is_type<v1::Reshape>(consumer)) {
+            for (const auto& in2 : consumer->output(0).get_target_inputs()) {
+                if (ov::is_type<v8::Softmax>(in2.get_node()))
```
[the remark this answers is not in the document]
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: updated

File: src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+#include "common_test_utils/ov_test_utils.hpp"
+
+#include "openvino/core/model.hpp"
+#include "openvino/op/add.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/parameter.hpp"
+#include "openvino/op/relu.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/result.hpp"
```
[the remark this answers is not in the document]
Comment (clee30) [reply src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp]: removed

File: src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+#include "openvino/op/select.hpp"
+#include "openvino/op/softmax.hpp"
+#include "openvino/pass/manager.hpp"
+#include "plugin/transformations/sdpa_select_mask_fusion.hpp"
+
+#include <limits>
```
[the remark this answers is not in the document]
Comment (clee30) [reply src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp:18]: still need for using numeric_limits

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            return false;
+
+        // The masked-out (else) value must be a scalar constant acting as -inf.
+        auto neg_inf_const = ov::as_type_ptr<v0::Constant>(pm.at(neg_inf).get_node_shared_ptr());
+        if (!neg_inf_const || ov::shape_size(neg_inf_const->get_shape()) != 1)
+            return false;
+        const float neg_inf_val = neg_inf_const->cast_vector<float>()[0];
+        // Accept values down to -1e4 to accommodate -inf saturated to -FP16_MAX (-65504).
+        if (neg_inf_val > -1e4f)
+            return false;
```
[the remark this answers is not in the document]
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: updated

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        // The rewrite is only equivalent once normalized by a following Softmax.
+        if (!feeds_softmax(pm.at(select_m)))
+            return false;
+
+        // additive_mask = Select(mask, 0, -inf); result = scores + additive_mask.
+        auto zero = v0::Constant::create(et, ov::Shape{}, {0.0f});
+        auto neg_inf_new = v0::Constant::create(et, ov::Shape{}, {-std::numeric_limits<float>::infinity()});
+        auto add_mask = std::make_shared<v1::Select>(cond_out, zero, neg_inf_new);
+        auto add = std::make_shared<v1::Add>(scores_out, add_mask);
```
Comment (allnes) [inline src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: The rewrite is not equivalent to the original `Select`. For `mask == false`, the original expression returns `negative_value`, while the replacement returns `scores + (-inf)`.

This changes the result for supported finite sentinels. With `scores=[-20000, 0]`, `mask=[true, false]`, and `negative_value=-10000`, the original graph produces `Softmax([-20000, -10000]) = [0, 1]`; the rewritten graph produces `Softmax([-20000, -inf]) = [1, 0]`. The measured maximum absolute error is `1`.

It is also incorrect for a true `-inf` sentinel when a masked score is NaN: `Select(false, NaN, -inf)` returns `-inf`, but `NaN + -inf` remains NaN and contaminates the Softmax row. A fully masked row with a finite sentinel changes from a uniform distribution to NaNs.

This conversion cannot be applied based only on the sentinel threshold and a downstream Softmax. Preserve boolean-mask semantics in the SDPA fusion/primitive, or require and enforce all conditions needed for equivalence. Add numerical GPU tests for finite sentinels, masked NaN/Inf values, and fully masked rows.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        if (neg_inf_val > kNegInfThreshold)
+            return false;
+
+        // The rewrite is only equivalent once normalized by a following Softmax.
+        if (!feeds_softmax(pm.at(select_m)))
+            return false;
+
+        // additive_mask = Select(mask, 0, -inf); result = scores + additive_mask.
+        auto zero = v0::Constant::create(et, ov::Shape{}, {0.0f});
+        auto neg_inf_new = v0::Constant::create(et, ov::Shape{}, {-std::numeric_limits<float>::infinity()});
```
Comment (allnes) [inline src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: `Constant::create(et, {}, {-std::numeric_limits<float>::infinity()})` throws when `et == f16` because the conversion fails the range check in `src/core/src/op/constant.cpp:189`:

```text
-inf assigned value out of range for f16
```

Reproducer: f16 Q/K/V tensors with shape `[1,8,128,64]` and

```text
MatMul(Q, K) -> Select(mask, scores, -65504) -> Softmax -> MatMul(probs, V)
```

`compile_model()` succeeds on parent `40aa1dd1716c` and fails on this commit in both Release and RelWithDebInfo builds. Construct the value using a type-safe representation if this rewrite remains after fixing its semantic issue. Add an f16 GPU compile-and-infer test for the complete attention graph.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        auto add_mask = std::make_shared<v1::Select>(cond_out, zero, neg_inf_new);
```
Comment (allnes) [inline src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: The replacement drops `select_node->get_auto_broadcast()`. The new `Select` and `Add` therefore use NUMPY broadcasting even when the source `Select` uses PDPD.

For `scores=[2,3,4,5]`, `condition=[3,4]`, and `PDPD(axis=1)`, the source graph is valid and compiles on the parent commit. This commit fails while inferring the new `Add`:

```text
Argument shapes are inconsistent.
```

Apply the original broadcast specification to the replacement graph, or explicitly broadcast the mask to the selected output shape and use `NONE` afterwards. Add coverage for PDPD with a lower-rank condition and for `NONE`.

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp [hunk 9]
```diff
@@ -841,6 +842,10 @@ void TransformationsPipeline::apply(std::shared_ptr<ov::Model> func) {
         // whose ConstantFolding folds the mask away when indices and depth are constants.
         manager.register_pass<ov::intel_gpu::DecomposeOneHotNonConstValues>();
 
+        // Convert decomposed-attention Select (where) masks into additive masks so the common
+        // ov::pass::SDPAFusion (registered inside the CommonOptimizations below) can fuse them.
+        manager.register_pass<ov::intel_gpu::SDPASelectMaskFusion>();
```
Comment (allnes) [inline src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp]: The pass is registered independently of `enable_sdpa_optimization`, and `feeds_softmax()` only checks `Select -> [Reshape] -> Softmax`. It does not verify the QK MatMul, the final MatMul with V, or that common `SDPAFusion` accepts the graph.

Consequently, `GPU_ENABLE_SDPA_OPTIMIZATION=false` still executes the rewrite. It also changes unrelated `Select -> Softmax` graphs from

```text
Select(select_gpu_ref) -> SoftMax(softmax_gpu_bf)
```

to

```text
Select(select_gpu_ref) -> Eltwise(eltwise_simple_vload8) -> SoftMax(softmax_gpu_bf)
```

Gate the pass with the SDPA optimization property and match the complete fusible attention pattern, or move this handling into `SDPAFusion`. Add property-off and non-SDPA negative tests.

File: src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        return std::make_shared<ov::Model>(ov::OutputVector{softmax}, ov::ParameterVector{scores, cond});
+    };
+    model = build();
+    manager.register_pass<SDPASelectMaskFusion>();
+    model_ref = build();
+}
+
+}  // namespace intel_gpu
+}  // namespace test
+}  // namespace ov
```
Comment (allnes) [inline src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp:286]: These tests only compare f32 graph structure. They do not run shape inference through the production pipeline, compile the graph for GPU, evaluate outputs, or verify final SDPA selection. Therefore all five tests pass despite the f16 compilation failure, PDPD failure, and numerical mismatches above.

Add GPU compile/inference cases for f16, finite and infinite sentinels, masked NaN/Inf, fully masked rows, PDPD broadcasting, disabled SDPA optimization, and a non-SDPA `Select -> Softmax` graph. For the positive case, also verify that the runtime graph selects the intended SDPA implementation.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp (hunk 6 above)
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: Agreed.  The original rewrite was not equivalent. The fix commit changes the approach: the additive value is no longer the sentinel (finite or -inf) but a finite lowest(et)/2, and only saturated sentinels (≤ -65504, i.e. -inf or its half-precision saturation) are rewritten — your [-20000, 0]/-10000 example now keeps its Select.   Some numerical GPU tests added.  Please review.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp (hunk 7 above)
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: Fixed: the pass no longer builds a -inf constant — the additive mask value is a finite lowest(et)/2, and the pass only rewrites f16/bf16 scores. Your f16 [1,8,128,64] example is now covered by a functional test.

File: src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp (hunk 8 above)
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations/sdpa_select_mask_fusion.cpp]: Fixed.  The PDPD unit tests added.

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp (hunk 9 above)
Comment (clee30) [reply src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp]: Fixed.  The registration is now gated by enable_sdpa_optimization,

File: src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp (hunk 10 above)
Comment (clee30) [reply src/plugins/intel_gpu/tests/unit/transformations/sdpa_select_mask_fusion_test.cpp:286]: Please review.  Added various functional tests to infer on actual pipeline.

**Comment by clee30** [comment]
Another model is using this 

Select(mask, -1, scores) 

Will need push a new update to support this.
